core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,274 @@
+     * {@link ByteBufAllocator}.
+     */
+    protected HttpDeframer(ByteBufAllocator alloc) {
+        requireNonNull(alloc, "alloc");
+        input = new ByteBufDeframerInput(alloc);
+    }
+
+    /**
+     * Converts the specified {@link HttpData} to {@link ByteBuf}.
+     */
+    protected ByteBuf convertToByteBuf(HttpData data) {
+        return data.byteBuf();
+    }
+
+    /**
+     * Decodes a stream of {@link HttpData}s to N objects.
+     * This method will be called whenever an {@link HttpData} is signaled from {@link Publisher}.
+     */
+    protected abstract void process(HttpDeframerInput in, HttpDeframerOutput<T> out);
+
+    private void process(HttpObject data) {
+        if (data instanceof HttpHeaders) {
+            final HttpHeaders headers = (HttpHeaders) data;
+
+            if (headers instanceof ResponseHeaders) {
+                final ResponseHeaders responseHeaders = (ResponseHeaders) headers;
+                if (responseHeaders.status().isInformational()) {
+                    processInformationalHeaders(responseHeaders, outputQueue::addLast);
+                    return;
+                }
+            }
+
+            if (!sawLeadingHeaders) {
+                sawLeadingHeaders = true;
+                processHeaders((HttpHeaders) data, outputQueue::addLast);
+            } else {
+                processTrailers((HttpHeaders) data, outputQueue::addLast);
+            }
+            return;
+        }
+
+        if (data instanceof HttpData) {
+            input.add(convertToByteBuf((HttpData) data));
+            process(input, outputQueue::addLast);
+            input.discardReadBytes();
+        }
+    }
+
+    /**
+     * Decodes an informational {@link ResponseHeaders} to N objects.
+     */
+    protected void processInformationalHeaders(ResponseHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Decodes an non-informational {@link HttpHeaders} to N objects.
+     */
+    protected void processHeaders(HttpHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Decodes a {@link HttpHeaders trailers} to N objects.
+     */
+    protected void processTrailers(HttpHeaders in, HttpDeframerOutput<T> out) {}
+
+    /**
+     * Invoked when a {@link Throwable} is raised while deframing.
+     */
+    protected void processOnError(Throwable cause) {}
+
+    @Override
+    final SubscriptionImpl subscribe(SubscriptionImpl subscription) {
+        final SubscriptionImpl subscriptionImpl = super.subscribe(subscription);
+        if (subscribedUpdater.compareAndSet(this, 0, 1)) {
+            eventLoop = subscription.executor();
+            deferredInit();
+        }
+        return subscriptionImpl;
+    }
+
+    private void deferredInit() {
+        final Subscription upstream = this.upstream;
+        if (upstream != null && subscribed != 0) {
+            if (initializedUpdater.compareAndSet(this, 0, 1)) {
+                if (demand() > 0) {
+                    upstream.request(1);
```

Comment:
The `demand` will be decreased when writing `deframed` data because this processor handles N to M conversion.
https://github.com/line/armeria/blob/fb091d30f5ad3b189f185bc744911dc59c176634/core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java#L209

Reply:
Ah, thanks. I missed that. 😅